### ISSUE TITLE
Dead code in leave_scope?

### ISSUE THREAD

**Issue by p5pRT** [Perl/perl5#12489]
Migrated from [rt.perl.org#115254](https://rt-archive.perl.org/perl5/Ticket/Display.html?id=115254) (status was 'resolved')

Searchable as RT115254$

**Comment by p5pRT** [comment]
### From @cpansprout

I do not understand this hunk from this commit&#8203;:

From&#8203;: Nick Ing\-Simmons \<nik@&#8203;tiuk\.ti\.com>
Date&#8203;: Wed\, 19 Dec 2001 10&#8203;:19&#8203;:40 \+0000 \(\+0000\)
Subject&#8203;: Fix READONLY hashes&#8203;:
X\-Git\-Tag&#8203;: perl\-5\.7\.3~27^2~161
X\-Git\-Url&#8203;: http&#8203;://perl5\.git\.perl\.org/perl\.git/commitdiff\_plain/a26e96dff1c9d91b0bf389e6149016d8a3f26e49?hp=1915677f475bd66f40974d770bf0b769f8cd7025

Fix READONLY hashes&#8203;:
 1\. Discovered need for disabled scope\.c part of Jeff's patch\.
 2\. Corrected hv\_delete\* \- entry was being de\-linked from
&nbsp;    chain \(patch program not smart enough?\)

p4raw\-id&#8203;: //depot/perlio@&#8203;13795


---
\.\.\. hv\.c diff snipped\.\.\.
<details><summary>Inline Patch</summary>

```diff
diff --git a/scope.c b/scope.c
index da5fa6b..7c02951 100644
--- a/scope.c
+++ b/scope.c
@@ -827,9 +827,8 @@ Perl_leave_scope(pTHX_ I32 base)
 		 * if a my variable that was made readonly is going out of
 		 * scope, we want to remove the readonlyness so that it can
 		 * go out of scope quietly
-		 * Disabled as I don't see need yet NI-S 2001/12/18
 		 */
-		if (0 && SvPADMY(sv) && ! SvFAKE(sv))
+		if (SvPADMY(sv) && !SvFAKE(sv))
 		    SvREADONLY_off(sv);
 
 		if (SvTHINKFIRST(sv))


```
</details>
That code does not turn off the read-only flag for read-only globs and regexps.  Yet that just works; no complaints about read-onliness:

$ \./perl \-Ilib \-we '\{my $x = $\{\+qr\}\}\}; Internals&#8203;::SvREADONLY\($x\,1\); \} print "ok\\n"'
ok
Pint&#8203;:perl\.git\-copy sprout$ \./perl \-Ilib \-we '\{my $x = \*foo; Internals&#8203;::SvREADONLY\($x\,1\); \} print "ok\\n"'
Name "main&#8203;::foo" used only once&#8203;: possible typo at \-e line 1\.
ok



---
Flags&#8203;:
&nbsp;    category=core
&nbsp;    severity=low


---
Site configuration information for perl 5\.17\.5&#8203;:

Configured by sprout at Sat Sep 22 18&#8203;:51&#8203;:23 PDT 2012\.

Summary of my perl5 \(revision 5 version 17 subversion 5\) configuration&#8203;:
&nbsp;  Snapshot of&#8203;: 451f421fe4742646fa2efbed0f45a19f0713d00f
&nbsp;  Platform&#8203;:
&nbsp;    osname=darwin\, osvers=10\.5\.0\, archname=darwin\-2level
&nbsp;    uname='darwin pint\.local 10\.5\.0 darwin kernel version 10\.5\.0&#8203;: fri nov 5 23&#8203;:20&#8203;:39 pdt 2010; root&#8203;:xnu\-1504\.9\.17~1release\_i386 i386 '
&nbsp;    config\_args='\-de \-Dusedevel \-DDEBUGGING'
&nbsp;    hint=recommended\, useposix=true\, d\_sigaction=define
&nbsp;    useithreads=undef\, usemultiplicity=undef
&nbsp;    useperlio=define\, d\_sfio=undef\, uselargefiles=define\, usesocks=undef
&nbsp;    use64bitint=undef\, use64bitall=undef\, uselongdouble=undef
&nbsp;    usemymalloc=n\, bincompat5005=undef
&nbsp;  Compiler&#8203;:
&nbsp;    cc='cc'\, ccflags ='\-fno\-common \-DPERL\_DARWIN \-DDEBUGGING \-fno\-strict\-aliasing \-pipe \-fstack\-protector \-I/usr/local/include'\,
&nbsp;    optimize='\-O3 \-g'\,
&nbsp;    cppflags='\-fno\-common \-DPERL\_DARWIN \-DDEBUGGING \-fno\-strict\-aliasing \-pipe \-fstack\-protector \-I/usr/local/include'
&nbsp;    ccversion=''\, gccversion='4\.2\.1 \(Apple Inc\. build 5664\)'\, gccosandvers=''
&nbsp;    intsize=4\, longsize=4\, ptrsize=4\, doublesize=8\, byteorder=1234
&nbsp;    d\_longlong=define\, longlongsize=8\, d\_longdbl=define\, longdblsize=16
&nbsp;    ivtype='long'\, ivsize=4\, nvtype='double'\, nvsize=8\, Off\_t='off\_t'\, lseeksize=8
&nbsp;    alignbytes=8\, prototype=define
&nbsp;  Linker and Libraries&#8203;:
&nbsp;    ld='env MACOSX\_DEPLOYMENT\_TARGET=10\.3 cc'\, ldflags =' \-fstack\-protector \-L/usr/local/lib'
&nbsp;    libpth=/usr/local/lib /usr/lib
&nbsp;    libs=\-ldbm \-ldl \-lm \-lutil \-lc
&nbsp;    perllibs=\-ldl \-lm \-lutil \-lc
&nbsp;    libc=\, so=dylib\, useshrplib=false\, libperl=libperl\.a
&nbsp;    gnulibc\_version=''
&nbsp;  Dynamic Linking&#8203;:
&nbsp;    dlsrc=dl\_dlopen\.xs\, dlext=bundle\, d\_dlsymun=undef\, ccdlflags=' '
&nbsp;    cccdlflags=' '\, lddlflags=' \-bundle \-undefined dynamic\_lookup \-L/usr/local/lib \-fstack\-protector'

Locally applied patches&#8203;:
&nbsp;    



---
@&#8203;INC for perl 5\.17\.5&#8203;:
&nbsp;    /usr/local/lib/perl5/site\_perl/5\.17\.5/darwin\-2level
&nbsp;    /usr/local/lib/perl5/site\_perl/5\.17\.5
&nbsp;    /usr/local/lib/perl5/5\.17\.5/darwin\-2level
&nbsp;    /usr/local/lib/perl5/5\.17\.5
&nbsp;    /usr/local/lib/perl5/site\_perl
&nbsp;    \.



---
Environment for perl 5\.17\.5&#8203;:
&nbsp;    DYLD\_LIBRARY\_PATH \(unset\)
&nbsp;    HOME=/Users/sprout
&nbsp;    LANG=en\_US\.UTF\-8
&nbsp;    LANGUAGE \(unset\)
&nbsp;    LD\_LIBRARY\_PATH \(unset\)
&nbsp;    LOGDIR \(unset\)
&nbsp;    PATH=/usr/bin&#8203;:/bin&#8203;:/usr/sbin&#8203;:/sbin&#8203;:/usr/local/bin&#8203;:/usr/X11/bin&#8203;:/usr/local/bin
&nbsp;    PERL\_BADLANG \(unset\)
&nbsp;    SHELL=/bin/bash

**Comment by p5pRT** [comment]
### From @rjbs

\* Father Chrysostomos \<perlbug\-followup@&#8203;perl\.org> \[2012\-10\-14T16&#8203;:12&#8203;:57\]
> $ \./perl \-Ilib \-we '\{my $x = $\{\+qr\}\}\}; Internals&#8203;::SvREADONLY\($x\,1\); \} print "ok\\n"'
> ok

The choice of qr\}\} there is \*really\* bizarre and\, I'd say\, unhelpful to
understanding the bug report without having to squint\.

Is there a technical reason to use this kind of construct instead of the one
that any report reader would understand?  I mean&#8203;:  it seems to cloud the issue
rather than demonstrate it in bare simplicity\.

\-\- 
rjbs

**Comment by p5pRT** [comment]
The RT System itself - Status changed from 'new' to 'open'

**Comment by p5pRT** [comment]
### From @cpansprout

On Sun Oct 14 13&#8203;:29&#8203;:34 2012\, perl\.p5p@&#8203;rjbs\.manxome\.org wrote&#8203;:
> \* Father Chrysostomos \<perlbug\-followup@&#8203;perl\.org> \[2012\-10\-
> 14T16&#8203;:12&#8203;:57\]
> > $ \./perl \-Ilib \-we '\{my $x = $\{\+qr\}\}\}; Internals&#8203;::SvREADONLY\($x\,1\);
> \} print "ok\\n"'
> > ok
> 
> The choice of qr\}\} there is \*really\* bizarre and\, I'd say\, unhelpful
> to
> understanding the bug report without having to squint\.
> 
> Is there a technical reason to use this kind of construct instead of
> the one
> that any report reader would understand?

No\.

> I mean&#8203;:  it seems to cloud
> the issue
> rather than demonstrate it in bare simplicity\.

True\.

I am not afraid of playing with edge cases when trying things out in
one\-liners\.  I often discover new tricks that way\.  But I forgot to
clean it up before submitting it\.  Sorry about that\.

\-\- 

Father Chrysostomos

**Comment by p5pRT** [comment]
### From [Unknown Contact. See original ticket]

On Sun Oct 14 13&#8203;:29&#8203;:34 2012\, perl\.p5p@&#8203;rjbs\.manxome\.org wrote&#8203;:
> \* Father Chrysostomos \<perlbug\-followup@&#8203;perl\.org> \[2012\-10\-
> 14T16&#8203;:12&#8203;:57\]
> > $ \./perl \-Ilib \-we '\{my $x = $\{\+qr\}\}\}; Internals&#8203;::SvREADONLY\($x\,1\);
> \} print "ok\\n"'
> > ok
> 
> The choice of qr\}\} there is \*really\* bizarre and\, I'd say\, unhelpful
> to
> understanding the bug report without having to squint\.
> 
> Is there a technical reason to use this kind of construct instead of
> the one
> that any report reader would understand?

No\.

> I mean&#8203;:  it seems to cloud
> the issue
> rather than demonstrate it in bare simplicity\.

True\.

I am not afraid of playing with edge cases when trying things out in
one\-liners\.  I often discover new tricks that way\.  But I forgot to
clean it up before submitting it\.  Sorry about that\.

\-\- 

Father Chrysostomos

**Comment by p5pRT** [comment]
### From @rjbs

\* Father Chrysostomos via RT \<perlbug\-comment@&#8203;perl\.org> \[2012\-10\-14T20&#8203;:46&#8203;:24\]
> I am not afraid of playing with edge cases when trying things out in
> one\-liners\.  I often discover new tricks that way\.  But I forgot to
> clean it up before submitting it\.  Sorry about that\.

For the record\, I do enjoy the weird\-o stuff I have seen you throw together\.
It's only when I see it in a bug report or similar place that I pout\.  So&#8203;:
please do not let me kill your overall sense of whimsy\. :\-\)

\-\- 
rjbs

**Comment by p5pRT** [comment]
### From @rurban

On Sun\, Oct 14\, 2012 at 3&#8203;:12 PM\, Father Chrysostomos
\<perlbug\-followup@&#8203;perl\.org> wrote&#8203;:
> I do not understand this hunk from this commit&#8203;:
>
> From&#8203;: Nick Ing\-Simmons \<nik@&#8203;tiuk\.ti\.com>
> Date&#8203;: Wed\, 19 Dec 2001 10&#8203;:19&#8203;:40 \+0000 \(\+0000\)
> Subject&#8203;: Fix READONLY hashes&#8203;:
> X\-Git\-Tag&#8203;: perl\-5\.7\.3~27^2~161
> X\-Git\-Url&#8203;: http&#8203;://perl5\.git\.perl\.org/perl\.git/commitdiff\_plain/a26e96dff1c9d91b0bf389e6149016d8a3f26e49?hp=1915677f475bd66f40974d770bf0b769f8cd7025
>
> Fix READONLY hashes&#8203;:
>  1\. Discovered need for disabled scope\.c part of Jeff's patch\.
>  2\. Corrected hv\_delete\* \- entry was being de\-linked from
>     chain \(patch program not smart enough?\)
>
> p4raw\-id&#8203;: //depot/perlio@&#8203;13795
> \-\-\-
>
> \.\.\. hv\.c diff snipped\.\.\.
> diff \-\-git a/scope\.c b/scope\.c
> index da5fa6b\.\.7c02951 100644
> \-\-\- a/scope\.c
> \+\+\+ b/scope\.c
> @&#8203;@&#8203; \-827\,9 \+827\,8 @&#8203;@&#8203; Perl\_leave\_scope\(pTHX\_ I32 base\)
>                  \* if a my variable that was made readonly is going out of
>                  \* scope\, we want to remove the readonlyness so that it can
>                  \* go out of scope quietly
> \-                \* Disabled as I don't see need yet NI\-S 2001/12/18
>                  \*/
> \-               if \(0 && SvPADMY\(sv\) && \! SvFAKE\(sv\)\)
> \+               if \(SvPADMY\(sv\) && \!SvFAKE\(sv\)\)
>                     SvREADONLY\_off\(sv\);
>
>                 if \(SvTHINKFIRST\(sv\)\)
>
> That code does not turn off the read\-only flag for read\-only globs and regexps\.  Yet that just works; no complaints about read\-onliness&#8203;:

He talks about readonly hashes\, not scalars\.
readonly hashes were different in its readonly\-ness \("restricted
hashes"\) than other SVs for some time\. Still are\.

But it looks like dead code indeed\.
I re\-added the 0 && and got the same results\. No readonly warnings
when clearing a HV or hvref pad\.

\./miniperl \-Ilib \-we '\{my%x=\(1\,0\);Internals&#8203;::SvREADONLY\(%h\,1\);\}print "ok\\n"'
\./miniperl \-Ilib \-we '\{my$x=\{1\,0\};Internals&#8203;::SvREADONLY\($x\,1\);\}print "ok\\n"'

I wonder until when a PL\_no\_modify warning appeared with hv\_clear\,
sv\_clear or av\_clear\.
hv\_name\_set\(\) for stashes maybe? But stashes cannot be lexicals yet\.

> $ \./perl \-Ilib \-we '\{my $x = $\{\+qr\}\}\}; Internals&#8203;::SvREADONLY\($x\,1\); \} print "ok\\n"'
> ok
> Pint&#8203;:perl\.git\-copy sprout$ \./perl \-Ilib \-we '\{my $x = \*foo; Internals&#8203;::SvREADONLY\($x\,1\); \} print "ok\\n"'
> Name "main&#8203;::foo" used only once&#8203;: possible typo at \-e line 1\.
> ok

We don't even have syntax yet to support readonly lexicals\, so I would
like to get the whole 2 lines
in question removed\. NI\-S 2001/12/18 seemed to be right\.

I cannot yet run the testsuite as somebody broke  make\(ext/DynaLoader\)\,
global\-buffer\-overflow\, probably a PL\_fold underflow in
S\_regmatch UCHARAT\(s\) \!= fold\_array\[nextchr\]\, nextchr being \-66368
\-\- 
Reini Urban
http&#8203;://cpanel\.net/   http&#8203;://www\.perl\-compiler\.org/

**Comment by p5pRT** [comment]
### From @rurban

On Wed\, Oct 17\, 2012 at 8&#8203;:18 PM\, Reini Urban \<rurban@&#8203;x\-ray\.at> wrote&#8203;:
> On Sun\, Oct 14\, 2012 at 3&#8203;:12 PM\, Father Chrysostomos
> \<perlbug\-followup@&#8203;perl\.org> wrote&#8203;:
>> I do not understand this hunk from this commit&#8203;:
>>
>> From&#8203;: Nick Ing\-Simmons \<nik@&#8203;tiuk\.ti\.com>
>> Date&#8203;: Wed\, 19 Dec 2001 10&#8203;:19&#8203;:40 \+0000 \(\+0000\)
>> Subject&#8203;: Fix READONLY hashes&#8203;:
>> X\-Git\-Tag&#8203;: perl\-5\.7\.3~27^2~161
>> X\-Git\-Url&#8203;: http&#8203;://perl5\.git\.perl\.org/perl\.git/commitdiff\_plain/a26e96dff1c9d91b0bf389e6149016d8a3f26e49?hp=1915677f475bd66f40974d770bf0b769f8cd7025
>>
>> Fix READONLY hashes&#8203;:
>>  1\. Discovered need for disabled scope\.c part of Jeff's patch\.
>>  2\. Corrected hv\_delete\* \- entry was being de\-linked from
>>     chain \(patch program not smart enough?\)
>>
>> p4raw\-id&#8203;: //depot/perlio@&#8203;13795
>> \-\-\-
>>
>> \.\.\. hv\.c diff snipped\.\.\.
>> diff \-\-git a/scope\.c b/scope\.c
>> index da5fa6b\.\.7c02951 100644
>> \-\-\- a/scope\.c
>> \+\+\+ b/scope\.c
>> @&#8203;@&#8203; \-827\,9 \+827\,8 @&#8203;@&#8203; Perl\_leave\_scope\(pTHX\_ I32 base\)
>>                  \* if a my variable that was made readonly is going out of
>>                  \* scope\, we want to remove the readonlyness so that it can
>>                  \* go out of scope quietly
>> \-                \* Disabled as I don't see need yet NI\-S 2001/12/18
>>                  \*/
>> \-               if \(0 && SvPADMY\(sv\) && \! SvFAKE\(sv\)\)
>> \+               if \(SvPADMY\(sv\) && \!SvFAKE\(sv\)\)
>>                     SvREADONLY\_off\(sv\);
>>
>>                 if \(SvTHINKFIRST\(sv\)\)
>>
>> That code does not turn off the read\-only flag for read\-only globs and regexps\.  Yet that just works; no complaints about read\-onliness&#8203;:
>
> He talks about readonly hashes\, not scalars\.
> readonly hashes were different in its readonly\-ness \("restricted
> hashes"\) than other SVs for some time\. Still are\.
>
> But it looks like dead code indeed\.
> I re\-added the 0 && and got the same results\. No readonly warnings
> when clearing a HV or hvref pad\.
>
> \./miniperl \-Ilib \-we '\{my%x=\(1\,0\);Internals&#8203;::SvREADONLY\(%h\,1\);\}print "ok\\n"'
> \./miniperl \-Ilib \-we '\{my$x=\{1\,0\};Internals&#8203;::SvREADONLY\($x\,1\);\}print "ok\\n"'
>
> I wonder until when a PL\_no\_modify warning appeared with hv\_clear\,
> sv\_clear or av\_clear\.
> hv\_name\_set\(\) for stashes maybe? But stashes cannot be lexicals yet\.
>
>> $ \./perl \-Ilib \-we '\{my $x = $\{\+qr\}\}\}; Internals&#8203;::SvREADONLY\($x\,1\); \} print "ok\\n"'
>> ok
>> Pint&#8203;:perl\.git\-copy sprout$ \./perl \-Ilib \-we '\{my $x = \*foo; Internals&#8203;::SvREADONLY\($x\,1\); \} print "ok\\n"'
>> Name "main&#8203;::foo" used only once&#8203;: possible typo at \-e line 1\.
>> ok
>
> We don't even have syntax yet to support readonly lexicals\, so I would
> like to get the whole 2 lines
> in question removed\. NI\-S 2001/12/18 seemed to be right\.

Nope\, he was not\.
There is one edge case remaining\, which prevents us from removing
those two lines\.

\./miniperl t/op/sort\.t
\.\.\.
ok 143 \- in\-place sort of read\-only array
Modification of a read\-only value attempted at t/op/sort\.t line 786\.
and it dies\, instead of continuing\.

\-\- 
Reini Urban
http&#8203;://cpanel\.net/   http&#8203;://www\.perl\-compiler\.org/

**Comment by p5pRT** [comment]
### From @cpansprout

On Thu Oct 18 07&#8203;:37&#8203;:46 2012\, rurban wrote&#8203;:
> There is one edge case remaining\, which prevents us from removing
> those two lines\.
> 
> \./miniperl t/op/sort\.t
> \.\.\.
> ok 143 \- in\-place sort of read\-only array
> Modification of a read\-only value attempted at t/op/sort\.t line 786\.
> and it dies\, instead of continuing\.

But\, still\, those two lines are wrong \(or the first one of them\)&#8203;:

$ \./perl \-Ilib \-e '\{ my $x = 3; Internals&#8203;::SvREADONLY $x\, 1; \(\) \}'
$ \./perl \-Ilib \-e '\{ my $x = $\{qr//\}; Internals&#8203;::SvREADONLY $x\, 1; \(\) \}'
Modification of a read\-only value attempted at \-e line 1\.

It may or may not die\, depending on the assigned value\.

\-\- 

Father Chrysostomos

**Comment by p5pRT** [comment]
### From @cpansprout

On Sat Sep 20 00&#8203;:19&#8203;:29 2014\, sprout wrote&#8203;:
> On Thu Oct 18 07&#8203;:37&#8203;:46 2012\, rurban wrote&#8203;:
> > There is one edge case remaining\, which prevents us from removing
> > those two lines\.
> > 
> > \./miniperl t/op/sort\.t
> > \.\.\.
> > ok 143 \- in\-place sort of read\-only array
> > Modification of a read\-only value attempted at t/op/sort\.t line 786\.
> > and it dies\, instead of continuing\.
> 
> But\, still\, those two lines are wrong \(or the first one of them\)&#8203;:
> 
> $ \./perl \-Ilib \-e '\{ my $x = 3; Internals&#8203;::SvREADONLY $x\, 1; \(\) \}'
> $ \./perl \-Ilib \-e '\{ my $x = $\{qr//\}; Internals&#8203;::SvREADONLY $x\, 1; \(\) \}'
> Modification of a read\-only value attempted at \-e line 1\.
> 
> It may or may not die\, depending on the assigned value\.

Fixed in 57c404c9c\.


\-\- 

Father Chrysostomos

**Comment by p5pRT** [comment]
### From [Unknown Contact. See original ticket]

On Sat Sep 20 00&#8203;:19&#8203;:29 2014\, sprout wrote&#8203;:
> On Thu Oct 18 07&#8203;:37&#8203;:46 2012\, rurban wrote&#8203;:
> > There is one edge case remaining\, which prevents us from removing
> > those two lines\.
> > 
> > \./miniperl t/op/sort\.t
> > \.\.\.
> > ok 143 \- in\-place sort of read\-only array
> > Modification of a read\-only value attempted at t/op/sort\.t line 786\.
> > and it dies\, instead of continuing\.
> 
> But\, still\, those two lines are wrong \(or the first one of them\)&#8203;:
> 
> $ \./perl \-Ilib \-e '\{ my $x = 3; Internals&#8203;::SvREADONLY $x\, 1; \(\) \}'
> $ \./perl \-Ilib \-e '\{ my $x = $\{qr//\}; Internals&#8203;::SvREADONLY $x\, 1; \(\) \}'
> Modification of a read\-only value attempted at \-e line 1\.
> 
> It may or may not die\, depending on the assigned value\.

Fixed in 57c404c9c\.


\-\- 

Father Chrysostomos

**Comment by p5pRT** [comment]
@cpansprout - Status changed from 'open' to 'resolved'
